Maintenance: create headless-git.exe to avoid foreground windows

derrickstolee opened the pull request:
Commits pulled from @dscho's derrickstolee/git#14.

This PR exists to more easily generate a Windows installer to check the whole flow.

dscho committed:
win32: add a helper to run `git.exe` without a foreground window

On Windows, there are two kinds of executables, console ones and
non-console ones. Git's executables are all console ones.

When launching the former e.g. in a scheduled task, a CMD window pops
up. This is not what we want for the tasks installed via the `git
maintenance` command.

To work around this, let's introduce `headless-git.exe`, which is a
non-console program that does _not_ pop up any window. All it does is to
re-launch `git.exe`, suppressing that console window, passing through
all command-line arguments as-are.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
git maintenance: avoid console window in scheduled tasks on Windows

We just introduced a helper to avoid showing a console window when the
scheduled task runs `git.exe`. Let's actually use it.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
Good idea! This shows two problems, one in the minimal Git artifact (I think), and the other in the vs-build part (where I did not make the necessary adjustments to build `headless-git.exe`).

dscho commented:
I'll continue playing with this later.

dscho commented:
@derrickstolee I'm happy to report that I got the build green (finally!).

derrickstolee committed:
fixup! maintenance: use Windows scheduled tasks

derrickstolee commented:
Needed to add 12fd7fa because an encoding mismatch caused `schtasks` failures on my machine. How did I not catch this before?

derrickstolee commented:
@jeffhostetler I could use your help here. I get the following BUG with my Scalar repos when running the maintenance this way:

```
BUG: gvfs-helper-client.c:189: verify_odb_line: unexpeced odb path 'C:/.scalarCache/id_ba47400f-eebe-4850-a8bf-ef694ef81414' vs 'C:\.scalarCache\id_ba47400f-eebe-4850-a8bf-ef694ef81414'
```

I'm sure it's related to [my commit making maintenance care about `gvfs.sharedCache`](https://github.com/microsoft/git/pull/301/commits/2864f7cbfd2901fb6a48f67260261ec0731cb5a2), and so I probably did it wrong. However, how can we make this more robust? Or, is this check actually helping us anymore? Was it just for testing?

_Edit: I added a90be59 as a potential workaround._

derrickstolee committed:
gvfs-helper-client: remove odb check

This check verifies that the ODB matches what we supplied, but there are
some subtleties around Windows path names that can cause inexact matches
to be logically the same. Since this check is really intended only for
debugging and development purposes, let's remove it for now as a quick
workaround.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

jeffhostetler commented:
i think it is fine to remove the check.  i shouldn't have used strcmp() there anyway. or rather, should have normalized probably both args first (to get the same type of slashes).

besides, the second BUG would have crashed it the structure arg was null.

thanks

derrickstolee commented:
Interesting that I can't re-target this PR after its initial target was deleted. I'll leave it closed until we have a final upstream version to merge in.

dscho commented:
> Interesting that I can't re-target this PR after its initial target was deleted

Ooops. I had not even realized that I closed this PR. 🤦

derrickstolee commented on gvfs-helper-client.c:
```diff
@@ -168,28 +168,6 @@ static int gh_client__send__objects_prefetch(struct child_process *process,
-
-	if (!gh_client__chosen_odb ||
-	    strcmp(v1_odb_path, gh_client__chosen_odb->path))
-		BUG("verify_odb_line: unexpeced odb path '%s' vs '%s'",
```
@jeffhostetler @dscho I hit this BUG during _one_ of the functional tests near the end of the [Scalar release build process](https://dev.azure.com/mseng/Scalar/_build/results?buildId=14099960&view=logs&j=de015ea9-9ec8-581a-12fc-2741c13e533e&t=9dce4694-c742-5e62-10d3-fa0894667d5b&l=247) so I'm going to cherry-pick this on top of `vfs-2.30.0`.
